### PR TITLE
Julia BRT Pipeline and Priority Map

### PR DESCRIPTION
**Pull request by gottacatchenall**
This pipeline (`./pipelines/julia_sdm.json`) runs on my machine, few notes before trying to run:

- This requires rebuilding the Docker image to update to Julia v1.9.3 with `docker compose build`
- ~Currently this doesn't mask out open-water, which should probably be fixed because the SDM assigns positive probability to some open-water regions, though it fits (or overfits?) much closer to the occurrence points~ this happens now

### CODE REVIEW

**Comment by jmlord** [comment]
About masking open water, could the BinaryLayer script be useful?

File: runners/julia-dockerfile
Comment (jmlord) [inline runners/julia-dockerfile:1]: FYI I have rebuilt the runner with you changes: https://github.com/GEO-BON/biab-2.0/actions/runs/6382068808
You can now use the compiled image.

File: scripts/SDM/julia_sdms/generateBackground.yml [hunk 1]
```diff
@@ -0,0 +1,28 @@
...
+    description: Dataframe, presence data.
+    type: text/tab-separated-values
+    example: "/output/data/getObservations/9f7d1cc148464cd0517e01c67af0ab5b/obs_data.tsv"
+  predictors:
+    label: predictors
+    description: layer names (predictors) as a list
+    type: image/tiff;application=geotiff[]
+    example: ["/output/data/loadFromStac/ea82148a2926d97acf85cea61a110194/bio1_242b2b01561981-01-01.tif","/output/data/loadFromStac/ea82148a2926d97acf85cea61a110194/bio2_243a4e7f541981-01-01.tif"]
+  buffer_distance:
+    label: buffer_distance
```
Comment (jmlord) [inline scripts/SDM/julia_sdms/generateBackground.yml:19]: should be human-readable

File: scripts/SDM/julia_sdms/fitBRT.yml [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+  sdm:
+    label: sdm
+    description: map of predicted occurrence probability
+    type: image/tiff;application=geotiff
+  sdm_uncertainty:
+    label: sdm uncertainty
+    description: map of relative uncertainty 
+    type: image/tiff;application=geotiff
+  fit_stats:
+    label: fit_stats
```
Comment (jmlord) [inline scripts/SDM/julia_sdms/fitBRT.yml:33]: should be human-readable

File: scripts/SDM/julia_sdms/generateBackground.yml [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    description: the minimum distance between any presence and any pseudoabsence in meters 
+    type: int
+    example: 50000
+outputs:
+  background:
+    label: background
+    description: TSV file containing with the coordinates background points.
```
Comment (jmlord) [inline scripts/SDM/julia_sdms/generateBackground.yml:26]: containing with?

File: scripts/SDM/julia_sdms/loadCHELSA.yml [hunk 4]
```diff
@@ -0,0 +1,27 @@
+script: loadCHELSA.jl
+name: Load CHELSA
+description: "This script creates an SDM and uncertainty map based on using Boosted Regression Trees (BRTs) using the package SpeciesDistributionToolkit.jl and EvoTrees.jl"
```
Comment (jmlord) [inline scripts/SDM/julia_sdms/loadCHELSA.yml:3]: Wrong description

File: scripts/SDM/julia_sdms/loadCHELSA.yml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+    identifier: https://orcid.org/0000-0002-6506-6487
+inputs:
+  bbox:
+    label: bbox
+    description: Vector of float, bbox coordinates of the extent in the order xmin, ymin, xmax, ymax in coordinates
+    type: float[]
+    example: [-86.75, 33.72,-52.21,63.27]
+  pca:
+    label: pca layers
+    description: Boolean, whether to PCA predictor layers or not
```
Comment (jmlord) [inline scripts/SDM/julia_sdms/loadCHELSA.yml:15]: Please unfold PCA for description.

File: scripts/SDM/julia_sdms/loadCHELSA.yml [hunk 6]
```diff
@@ -0,0 +1,27 @@
+script: loadCHELSA.jl
```
Comment (jmlord) [inline scripts/SDM/julia_sdms/loadCHELSA.yml:1]: I'm not sure why you can't use CHELSA from the STAC, using loadFromStac?
https://coleo.biodiversite-quebec.ca/apps/io-layers/chelsa-clim/bio16
Do you need this to be in a specific format afterwards?

File: pipelines/julia_sdm.json [hunk 7]
```diff
@@ -0,0 +1,444 @@
...
+      }
+    ],
+    "inputs": {
+      "data>pyLoadObservations>pyLoadObservations.yml@96|data_source": {
+        "description": "Source of the data (One of gbif_pc - Planetary computer or gbif_api - GBIF Download API)",
+        "label": "Data source",
+        "type": "options",
+        "options": [
+          "gbif_pc",
+          "gbif_api"
```
Comment (jmlord) [inline pipelines/julia_sdm.json:352]: You should put gbif_api as a constant for now, latest news there was a bug with the PC and we were waiting for a fix from them.

**Review by jmlord — COMMENTED** [review]
Ok on my side with these minor changes.
We'd need an SDM person to review now...

**Comment by glaroc** [comment]
I think it would be important to have spatial resolution as a pipeline input for all SDMs.

**Comment by gottacatchenall** [comment]
> I think it would be important to have spatial resolution as a pipeline input for all SDMs.

I agree, and this also further highlights why we need a pipeline that is only about validating inputs

**Comment by glaroc** [comment]
On my work laptop, if I try to run the pipeline, I get 

````
[7] signal (4.2): Illegal instruction
in expression starting at /scripts/SDM/julia_sdms/loadCHELSA.jl:92
````

I did a docker compose build and pull. Any idea what is causing this ?

**Comment by gottacatchenall** [comment]
> On my work laptop, if I try to run the pipeline, I get
> 
> ```
> [7] signal (4.2): Illegal instruction
> in expression starting at /scripts/SDM/julia_sdms/loadCHELSA.jl:92
> ```
> 
> I did a docker compose build and pull. Any idea what is causing this ?

Yeah, JM was having that issue too and I'm not sure why, but regardless now that we can use `loadFromSTAC` again, we don't need to `loadCHELSA` script. I'll push an updated version soon

**Comment by gottacatchenall** [comment]
Okay, just pushed a version that works on my machine w/ loadFromSTAC

**Comment by frousseu** [comment]
I think there is something wrong with the pipeline json? 
image

**Comment by gottacatchenall** [comment]
That looks normal, there are still some unnecessary inputs on loadFromSTAC that aren't used but it should run as is

**Comment by frousseu** [comment]
Ok, it just looked very different from the previous one. I'm getting this error when trying to run the pipeline 
/scripts/SDM/julia_sdms/pcaLayers.yml (No such file or directory)

**Comment by gottacatchenall** [comment]
Oh I forgot to add the pcaLayers script files in git, I've just pushed them

**Comment by frousseu** [comment]
Have you considered creating the background using the gbif density rasters to account for the sampling bias? I'd be curious to see what it would look like especially vs. the uncertainty.

**Comment by frousseu** [comment]
I have this error from GenerateBackground:
```
[18] signal (4.2): Illegal instruction
in expression starting at /scripts/SDM/julia_sdms/generateBackground.jl:34
```
Let me know if you need more info. Also, I think there are to input areas for loadFromSTAC (and maybe both are running)

**Comment by gottacatchenall** [comment]
If there's any more to the error message that could help, line 34 is just the `main()` call so I think there should be more to the stacktrace.

**Comment by glaroc** [comment]
@gottacatchenall This is what I get

Invalid instruction at 0x7f69875ca089: 0x62, 0xf1, 0xf7, 0x08, 0x7b, 0x45, 0xf7, 0x48, 0x8b, 0x0d, 0x69, 0x48, 0x83, 0x01, 0x48

[21] signal (4.2): Illegal instruction
in expression starting at /scripts/SDM/julia_sdms/generateBackground.jl:34

**Comment by gottacatchenall** [comment]
Okay I think we fixed it, looks like not explicitly telling `CSV.read` that it was tab-delimited fails on some machines but not mine (idk why, kind of weird in docker)

**Comment by glaroc** [comment]
I'm still getting the same error.

**Comment by gottacatchenall** [comment]
@tpoi any thoughts on the above error? something weird with CSV reading tab delimited files

**Comment by tpoisot** [comment]
Do you have a copy of the CSV? I usually go for CSV.File instead of CSV.read, if that helps.

**Comment by gottacatchenall** [comment]
It would be best to test this with a TSV produced on a failed instance. Could either @glaroc @frousseu provide an example output of TSV of the `pyLoadObervations`->`cleanCoordinates` from their machine so I can check if reads normally w/ `CSV.File`

I'm still pretty baffled why a docker container would have different behaviors on different computers (docker version differences? idk)

**Comment by glaroc** [comment]
In fact, the pipeline runs with the default inputs you provided. However, when I choose other layers than CHELSA, for example elevation, earthenv land cover, etc. That's where I get the error.

**Comment by gottacatchenall** [comment]
Hexadecimals getting written as string literals? Idk

**Comment by tpoisot** [comment]
Could it _maybe_ be because the stream isn't read entirely? The only time I struggled with reading CSVs it was something like this (files corrupted during download)

**Comment by glaroc** [comment]
@gottacatchenall I sent you an example file in Slack. Github doesn't allow .tsv

**Comment by gottacatchenall** [comment]
I'm able to load it fine into Julia, so I don't think its corrupt

**Comment by gottacatchenall** [comment]
> Have you considered creating the background using the gbif density rasters to account for the sampling bias? I'd be curious to see what it would look like especially vs. the uncertainty.

we should 100% build a pipeline for this to compare the predictions with correction vs. without

**Comment by glaroc** [comment]
@gottacatchenall In the .json file created by Julia for fitBRT, the uncertainty layer is named 'uncertainty' instead of sdm_uncertainty. Can you please fix that ?

File: scripts/SDM/julia_sdms/loadCHELSA.yml (hunk 6 above)
Comment (gottacatchenall) [reply scripts/SDM/julia_sdms/loadCHELSA.yml:1]: no longer exists

File: pipelines/julia_sdm.json [hunk 8]
```diff
@@ -0,0 +1,444 @@
...
+        "example": 2010
+      },
+      "data>pyLoadObservations>pyLoadObservations.yml@96|max_year": {
+        "description": "Max year observations wanted",
+        "label": "maximum year",
+        "type": "int",
+        "example": 2020
+      },
+      "SDM>julia_sdms>loadCHELSA.yml@139|pca": {
+        "description": "Boolean, whether to PCA predictor layers or not",
```
[the remark this answers is not in the document]
Comment (gottacatchenall) [reply pipelines/julia_sdm.json]: no longer exists

File: pipelines/julia_sdm.json [hunk 9]
```diff
@@ -0,0 +1,444 @@
...
+        "label": "Fit Stats",
+        "type": "text/json"
+      },
+      "filtering>cleanCoordinates.yml@34|clean_presence": {
+        "description": "Data frame with clean occurrence points.",
+        "label": "Cleaned Occurrences",
+        "type": "text/tab-separated-values"
+      },
+      "SDM>julia_sdms>generateBackground.yml@125|background": {
+        "description": "TSV file containing with the background (pseudoabsence) coordinates.",
```
[the remark this answers is not in the document]
Comment (gottacatchenall) [reply pipelines/julia_sdm.json]: fixed in script yaml

File: pipelines/julia_sdm.json [hunk 10]
```diff
@@ -0,0 +1,444 @@
...
+        "example": 2020
+      },
+      "SDM>julia_sdms>loadCHELSA.yml@139|pca": {
+        "description": "Boolean, whether to PCA predictor layers or not",
+        "label": "pca layers",
+        "type": "boolean",
+        "example": true
+      },
+      "SDM>julia_sdms>loadCHELSA.yml@139|layer_numbers": {
+        "description": "The numbers of CHELSA layers to use",
```
[the remark this answers is not in the document]
Comment (gottacatchenall) [reply pipelines/julia_sdm.json]: no longer exists

File: scripts/SDM/julia_sdms/loadCHELSA.yml (hunk 5 above)
Comment (gottacatchenall) [reply scripts/SDM/julia_sdms/loadCHELSA.yml:15]: no longer exists

File: scripts/SDM/julia_sdms/generateBackground.yml (hunk 3 above)
Comment (gottacatchenall) [reply scripts/SDM/julia_sdms/generateBackground.yml:26]: fixed

**Comment by gottacatchenall** [comment]
@jmlord @glaroc  I have pushed a new pipeline `brt_into_priority_map.json`, which runs on my machine with the follow yml input:
```
SDM>BRT_julia_viewer.json@21|data>loadFromStac.yml@0|mask: null
SDM>BRT_julia_viewer.json@21|data>loadFromStac.yml@18|mask: null
pipeline@12: 2010
pipeline@15: 5
pipeline@16:
  - 0.3
  - 0.2
  - 0.4
  - 0.1
pipeline@17: 2020
pipeline@18:
  - Acer saccharum
pipeline@19: 150000
pipeline@22: 2010
```

and produces a priority map
image

**Comment by glaroc** [comment]
I tried running it on the server and I think we need to add the Clustering package

**Comment by gottacatchenall** [comment]
Yeah it requires a `docker build`

**Comment by glaroc** [comment]
I tried to change the pipeline to get more meaningful outputs. See : https://biab2.geobon.org/pipeline-form/Block3%3Ebrt_into_priority_map/69ffbd6e95a61471852cfbc1753b8063 

But climate velocity is returning an error. 

ERROR: LoadError: DimensionMismatch: arrays could not be broadcast to a common size; got a dimension with lengths 19 and 6

**Comment by gottacatchenall** [comment]
Do you have the line number?

**Comment by gottacatchenall** [comment]
The baseline layers passed to climateVelocity have to be correspond exactly to the projected variables, and the 19 chelsa variables are hardcoded as the future to compare to

**Comment by glaroc** [comment]
Is there is no way to include elevation and other variables in the SDM ?

I think in the future, it would make sense to put climate velocity and uniqueness in a separate script to clarify things.

**Comment by gottacatchenall** [comment]
I just pushed a version where uniqueness and the SDM work based on the inputted layers, and velocity is only calculated with the 19 bioclim vars for both the baseline and future

**Comment by glaroc** [comment]
@gottacatchenall Did you try running it ? This gives me a loadFromStac error: bbox
  Bounding box must be within (-180, -90, 180, 90) (type=value_error)

for one of the calls. Others seem ok.

**Comment by gottacatchenall** [comment]
I was running it with
```
Block3>block3.json@26|pipeline@3:
  - -86.748047
  - 33.72434
  - -52.207031
  - 63.273182
pipeline@12: 1995
pipeline@14:
  - chelsa-clim|bio1
  - chelsa-clim|bio7
  - chelsa-clim|bio12
  - earthenv_topography|elevation_median
  - earthenv_landcover|class_3
  - earthenv_landcover|class_12
pipeline@15: 5
pipeline@16:
  - 0.3
  - 0.2
  - 0.4
  - 0.1
pipeline@17: 2020
pipeline@18:
  - Acer saccharum
pipeline@19: 150000
pipeline@22: 1995
```

but the SDM is much better with all CHELSA layers

**Comment by jmlord** [comment]
Comment from meeting: Cleanup what are inputs and what are not inputs.

**Comment by frousseu** [comment]
The BRT pipeline can now be run on my Windows computer. For now, the two main things to improve would be:
1 - The ability to choose the species and the predictors
2 - There are duplicated inputs in the Input form (mask, weight matrix) and it is not clear why the weight matrix is there. Shouldn't it be in the pipeline for the priority map?
Otherwise, it would be nice to connect the julia brt with the createBackground function to get a biased background.

**Comment by frousseu** [comment]
Otherwise, I'm getting this error from the block3.json in the priority map pipeline:
```r
$ids
[1] "esacci-lc-2020"

Error: HTTP status '400'. Invalid parameters provided 1 validation error for SearchPostRequest
bbox
  Bounding box must be within (-180, -90, 180, 90) (type=value_error)
No traceback available 
Parent job is Cancelling: killing server process...
Parent job is Cancelling: done.
```

**Comment by gottacatchenall** [comment]
> Otherwise, I'm getting this error from the block3.json in the priority map pipeline:
> 
> ```r
> $ids
> [1] "esacci-lc-2020"
> 
> Error: HTTP status '400'. Invalid parameters provided 1 validation error for SearchPostRequest
> bbox
>   Bounding box must be within (-180, -90, 180, 90) (type=value_error)
> No traceback available 
> Parent job is Cancelling: killing server process...
> Parent job is Cancelling: done.
> ```

This is because the bounding box has to in long/lat coordinates for Julia right now as everything is in WGS84 (between -180 and 180 long and -85,85 lat)

**Comment by jmlord** [comment]
> The BRT pipeline can now be run on my Windows computer. For now, the two main things to improve would be: 1 - The ability to choose the species and the predictors 2 - There are duplicated inputs in the Input form (mask, weight matrix) and it is not clear why the weight matrix is there. Shouldn't it be in the pipeline for the priority map? Otherwise, it would be nice to connect the julia brt with the createBackground function to get a biased background.

I think Guillaume's recent modifications should have fixed these.

File: scripts/SDM/julia_sdms/generateBackground.jl
Comment (jmlord) [inline scripts/SDM/julia_sdms/generateBackground.jl:1]: More logs too, if relevant. Or maybe I'm not receiving the Julia logs in the output?
image

File: scripts/Block3/climateVelocity.yml
Comment (jmlord) [inline scripts/Block3/climateVelocity.yml:1]: Should be in the folder along with the other climate layers. There should be an explanation differentiating it with Juan's work.

File: scripts/SDM/julia_sdms/generateBackground.yml
Comment (jmlord) [inline scripts/SDM/julia_sdms/generateBackground.yml:1]: Any reason not to use the R version?

File: scripts/SDM/julia_sdms/fitBRT.yml [hunk 11]
```diff
@@ -0,0 +1,39 @@
...
+    label: sdm
+    description: map of predicted occurrence probability
+    type: image/tiff;application=geotiff
+  sdm_uncertainty:
+    label: sdm_uncertainty
+    description: map of relative uncertainty 
+    type: image/tiff;application=geotiff
+  fit_stats:
+    label: fit_stats
+    description: JSON of model fit stats and threshold
```
Comment (jmlord) [inline scripts/SDM/julia_sdms/fitBRT.yml:38]: Guidance for interpretation of the fit statistics would be appreciated.

File: scripts/SDM/julia_sdms/fitBRT.yml [hunk 12]
```diff
@@ -0,0 +1,39 @@
...
+        "/output/data/loadFromStac/ea82148a2926d97acf85cea61a110194/bio2_243a4e7f541981-01-01.tif",
+      ]
+outputs:
+  sdm:
+    label: sdm
+    description: map of predicted occurrence probability
+    type: image/tiff;application=geotiff
+  sdm_uncertainty:
+    label: sdm_uncertainty
+    description: map of relative uncertainty
```
Comment (jmlord) [inline scripts/SDM/julia_sdms/fitBRT.yml:34]: Guidance for interpretation: 1 uncertain and 0 certain?

File: scripts/Block3/weightLayers.jl
Comment (jmlord) [inline scripts/Block3/weightLayers.jl:1]: When I ran this, the output priority map was empty.

**Review by jmlord — CHANGES_REQUESTED** [review]
@gottacatchenall  A few clarifications in documentation please before we merge.

Also, most important: the priority map was empty when I ran it.

File: scripts/SDM/julia_sdms/generateBackground.jl
Comment (jmlord) [reply scripts/SDM/julia_sdms/generateBackground.jl:1]: Ok, I checked with the helloJulia example, and I got the logs.

**Comment by jmlord** [comment]
@gottacatchenall Can you investigate why this produces no result?
See also documentation clarification requests in review above.

**Comment by gottacatchenall** [comment]
Once https://github.com/PoisotLab/SpeciesDistributionToolkit.jl/pull/227 is merged, I will refactor this to be merged.

**Comment by jmlord** [comment]
Hi folks,
Any update on this?
Laura's lab intends to integrate a new bias correction algorithm and could use a BRT script in their pipeline.

**Comment by tpoisot** [comment]
I'm close but no ETA yet -- if all goes well, by the end of June? I'd need to see with @gottacatchenall how much work is needed to port his existing pipeline

**Comment by glaroc** [comment]
@gottacatchenall Do you have a description somewhere of the procedure you use to obtain the uncertainty layer for the BRT ? Thanks.

**Comment by gottacatchenall** [comment]
I've added a description of the BRT methodology (including how uncertainty is computed) [here](https://github.com/GEO-BON/bon-in-a-box-pipelines/blob/mdc/julia_brt/scripts/SDM/julia_sdms/README.md) .

A few things I noticed on revisiting this:

- Crossvalidation splits are included in the BRT script, which isn't ideal because we would like different CV methods to be different scripts that can be used an inputs to different SDMs
- This probably necessitates a subpipeline for repeated CV on a given model using a single CV method
- Computation of fit metrics is inevitably going to be duplicated so should be moved somewhere else

**Comment by gottacatchenall** [comment]
replaced with #170

**Comment by tpoisot** [comment]
image

**Pull request closed without merging**